horace_core/sqw/@sqw/func_eval.m

```diff
@@ -116,10 +129,175 @@
+        error( ...
+            'SQW:func_eval:invalid_argument', ...
+            ['Parameter ''filebacked'' must be a scalar logical value.\n', ...
+             'Found ''%s'' with size [%s].'], ...
+            class(keyval.filebacked), num2str(size(keyval.filebacked)) ...
+        );
+    end
+    opts.filebacked = keyval.filebacked;
+
+    if isempty(keyval.outfile)
+        opts.outfile = {};
+    elseif ~iscell(keyval.outfile)
+        opts.outfile = {keyval.outfile};
+    else
+        opts.outfile = keyval.outfile;
+    end
+
+    outfiles_empty = all(cellfun(@(x) isempty(x), opts.outfile));
+    if ~outfiles_empty && (numel(win) ~= numel(opts.outfile))
+        error( ...
+            'SQW:func_eval:invalid_arguments', ...
+            ['Number of outfiles specified must match number of input objects.\n' ...
+             'Found %i outfile(s), but %i sqw object(s).'], ...
+            numel(opts.outfile), numel(win) ...
+        );
+    end
+    if outfiles_empty && opts.filebacked
+        opts.outfile = gen_array_of_tmp_file_paths( ...
+            numel(win), 'horace_func_eval', tmp_dir() ...
+        );
+    end
+end
+
+
+function paths = gen_array_of_tmp_file_paths(nfiles, prefix, base_dir)
+    % Generate a cell array of paths for temporary files to be written to
+    % Format of the file names follows:
+    %   <prefix>_<UUID>_<counter_with_padded_zeros>.tmp
+    if nfiles < 1
+        error('CUT:cut_accumulate_data_', ...
+              ['Cannot create temporary file paths for less than 1 file.' ...
+               '\nFound %i.'], nfiles);
+    end
+    uuid = char(java.util.UUID.randomUUID());
+    counter_padding = floor(log10(nfiles)) + 1;
+    format_str = sprintf('%s_%s_%%0%ii.tmp', prefix, uuid, counter_padding);
+    paths = cell(1, nfiles);
+    for i = 1:nfiles
+        paths{i} = fullfile(base_dir, sprintf(format_str, i));
+    end
+end
+
+
+function write_sqw_with_out_of_mem_pix(sqw_obj, outfile)
+    % Write the given SQW object to the given file.
+    % The pixels of the SQW object will be derived from the image signal array
+    % and npix array, saving in chunks so they do not need to be held in memory.
+    %
+    pix = sqw_obj.data.pix;
+    npix = sqw_obj.data.npix;
+    img_signal = sqw_obj.data.s;
+
+    ldr = sqw_formats_factory.instance().get_pref_access(sqw_obj);
+    ldr = ldr.init(sqw_obj, outfile);
+    ldr.put_sqw('-nopix');
+    ldr = write_out_of_mem_pix(pix, npix, img_signal, ldr);
+    ldr = ldr.validate_pixel_positions();
+    ldr = ldr.put_footers();
+    ldr.delete();
+end
+
+
+function loader = write_out_of_mem_pix(pix, npix, img_signal, loader)
+    % Smear the given image signal values over a file-backed PixelData object
+    % Set each pixel's signal to the value of the average signal of the bin
+    % that pixel belongs to. Set all variances to zero.
+    %
+    pix.move_to_first_page();
+    npix_cum_sum = cumsum(npix(:));
+    end_idx = 1;
+    while true
```

Comment:
this logic is duplicated from other memory page to file writing. Is there any way to remove the duplication? This is quite a subtle piece of indexing.

Reply:
Issue opened here https://github.com/pace-neutrons/Horace/issues/592